### PR TITLE
INFINITY-2328 Log both package and SDK version when service is starting up

### PR DESCRIPTION
**Pull request by nickbp**
- SDK version info is automatically generated in the Java code (static members in a `SDKBuildInfo` class) when a Gradle build is run.
- Package version info is passed via Marathon envvars. This information is populated when the stub-universe is created (and updated by the release builder as well). Turns out that we can't access package.json content from marathon.json but no matter.

Result:
```
INFO  2017-09-06 23:51:07,891 [main] com.mesosphere.sdk.specification.DefaultService:<init>(84): Build information:
- hello-world: stub-universe, built 2017-09-06T23:48:49.702Z
- SDK: 0.21.0-SNAPSHOT/f132dec, built 2017-09-06T23:48:24.750Z
```

### CODE REVIEW

File: tools/release_builder.py [hunk 1]
```diff
@@ -491,10 +490,37 @@ def _update_name_and_version(self, pkgdir):
...
+        '''Updates the marathon.json definition to contain the desired name and version strings.
+        '''
+        # note: the file isn't valid JSON, so we edit the raw content instead
+        marathon_file_name = os.path.join(pkgdir, 'marathon.json.mustache')
+        with open(marathon_file_name) as f:
+            orig_marathon_lines = f.readlines()
+
+        marathon_lines = []
+        for line in orig_marathon_lines:
+            name_match = re.match(r'^ *"PACKAGE_NAME": ?"(.*)",?$', line.rstrip('\n'))
```
Comment (elezar) [inline tools/release_builder.py]: We could use `re.sub()` to perform the substitution directly. Also, is the `rstrip()` on the line strictly required?

File: tools/release_builder.py [hunk 2]
```diff
@@ -480,7 +479,7 @@ def _update_name_and_version(self, pkgdir):
         # Update package's version to reflect the user's input
         package_json['version'] = self._pkg_version
 
-        logger.info('[1/2] Updated package.json:')
+        logger.info('[1/3] Updated package.json:')
```
Comment (elezar) [inline tools/release_builder.py]: Seems almost like we could use something like a `log_progress` method where we track the current step?

File: frameworks/cassandra/universe/marathon.json.mustache [hunk 3]
```diff
@@ -22,6 +22,9 @@
   },
   {{/service.service_account_secret}}
   "env": {
+    "PACKAGE_NAME": "{{package-name}}",
+    "PACKAGE_VERSION": "{{package-version}}",
+    "PACKAGE_BUILD_TIME_MS": "{{package-build-time-ms}}",
```
[the remark this answers is not in the document]
Comment (nickbp) [reply frameworks/cassandra/universe/marathon.json.mustache]: - I'd picture "timestamp" as possibly implying a string value, while this is unix epoch time in milliseconds. `BUILD_TIME_MS` is trying to get across the format/units being used.
- I'd actually tried using `package.name` before, but in the context of having cosmos extract the `name` field from `package.json`, similar to e.g. `resource.foo.bar` extracting the `foo { bar: ... }` field from `resource.json`. However Cosmos apparently doesn't do this, hence this hack. However if they did someday, I could see there being a collision with use of `package.X`.

File: sdk/common/build.gradle [hunk 4]
```diff
@@ -18,6 +20,25 @@ task sourceJar(type: Jar) {
     from sourceSets.main.allJava
 }
 
+// Configure the SDKBuildInfo class generator
+def getGitHash = { ->
+    def stdout = new ByteArrayOutputStream()
+    exec {
+        commandLine 'git', 'rev-parse', '--short', 'HEAD'
```
[the remark this answers is not in the document]
Comment (nickbp) [reply sdk/common/build.gradle]: Oh yeah that'd be great. I'll see if I can find a way to do that here.

File: sdk/common/build.gradle [hunk 5]
```diff
@@ -18,6 +20,25 @@ task sourceJar(type: Jar) {
...
+    return stdout.toString().trim()
+}
+buildConfig {
+    packageName = 'com.mesosphere.sdk.generated'
+    clsName = 'SDKBuildInfo'
+
+    version = project.version // VERSION
+    appName = project.group + '.' + project.name // NAME
+    buildConfigField 'String', 'GIT_SHA', "${getGitHash()}"
+    buildConfigField 'long', 'BUILD_TIME_MS', System.currentTimeMillis() + 'L'
```
[the remark this answers is not in the document]
Comment (nickbp) [reply sdk/common/build.gradle]: epoch time in milliseconds

File: tools/release_builder.py (hunk 1 above)
Comment (nickbp) [reply tools/release_builder.py]: `line` ends with a trailing newline. I've had wonky behavior with passing newlines through regexes before.

File: tools/release_builder.py (hunk 2 above)
Comment (nickbp) [reply tools/release_builder.py]: I'll just remove the `[x/y]`s as they aren't really providing anything substantive

**Comment by gabrielhartmann** [comment]
@nickbp: Please file a JIRA for inability to use `package.json` in `marathon.json.mustache`.

File: frameworks/cassandra/universe/marathon.json.mustache (hunk 3 above)
[the remark this answers is not in the document]
Comment (nickbp) [reply frameworks/cassandra/universe/marathon.json.mustache]: (note lack of `package.json` mention in the bullets listed here: https://github.com/mesosphere/universe#marathonjsonmustache )

[the remark this answers is not in the document]
Comment (gabrielhartmann) [reply frameworks/cassandra/universe/marathon.json.mustache]: Can we change `PACKAGE_BUILD_TIME_MS` to `PACKAGE_BUILD_TIME_EPOCH_MS` or something, or perhaps we can put a more human readable timestamp indicating its in UTC there?

**Comment by nickbp** [comment]
Done: DCOS-18470

File: frameworks/cassandra/universe/marathon.json.mustache (hunk 3 above)
[the remark this answers is not in the document]
Comment (nickbp) [reply frameworks/cassandra/universe/marathon.json.mustache]: I'd really like to have a(n easily) machine-readable version in there. I'll add a second string-formatted value for debug access. Will also add a string representation to the SDK version info being autogenerated by gradle.

**Comment by benclarkwood** [comment]
[Add version and commit](https://trello.com/c/ZxcXSWXU/11-add-version-and-commit)

File: tools/release_builder.py (hunk 1 above)
Comment (nickbp) [reply tools/release_builder.py]: Tried `re.sub()`:
- Pattern to be used is significantly more spaghetti-like: `re.sub(r'^( *"{}": *")(.*)(",?)$'.format(name), r'\1{}\3'.format(val), line)`
- Regexp apparently doesn't work on certain input, for reasons that I can't figure out:
```
input: '    "PACKAGE_VERSION": "stub-universe",
'
Traceback (most recent call last):
  File "/usr/lib/python3.5/sre_parse.py", line 938, in expand_template
    literals[index] = g(group) or empty
IndexError: no such group

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "./tools/release_builder.py", line 583, in <module>
    sys.exit(main(sys.argv))
  File "./tools/release_builder.py", line 564, in main
    response = builder.release_package()
  File "./tools/release_builder.py", line 526, in release_package
    self._update_marathon_json(pkgdir)
  File "./tools/release_builder.py", line 511, in _update_marathon_json
    line = with_updated_envvar(line, "PACKAGE_VERSION", self._pkg_version)
  File "./tools/release_builder.py", line 506, in with_updated_envvar
    newline = re.sub(r'^( *"{}": *")(.*)(",?)$'.format(name), r'\1{}\3'.format(val), line)
  File "/usr/lib/python3.5/re.py", line 182, in sub
    return _compile(pattern, flags).sub(repl, string, count)
  File "/usr/lib/python3.5/re.py", line 330, in filter
    return sre_parse.expand_template(template, match)
  File "/usr/lib/python3.5/sre_parse.py", line 940, in expand_template
    raise error("invalid group reference")
sre_constants.error: invalid group reference
```

Going to stick with `re.match` + `str.replace` until someone who's better at regexes can figure out the correct permutation to use here.

File: frameworks/cassandra/universe/marathon.json.mustache (hunk 3 above)
[the remark this answers is not in the document]
Comment (nickbp) [reply frameworks/cassandra/universe/marathon.json.mustache]: Both machine-readable epoch-ms and user-facing string timestamp values are now present

File: sdk/common/build.gradle (hunk 4 above)
[the remark this answers is not in the document]
Comment (nickbp) [reply sdk/common/build.gradle]: Done! `-dirty` should be added to the SHA when relevant

File: tools/release_builder.py (hunk 1 above)
Comment (nickbp) [reply tools/release_builder.py]: Yes, marathon.json.mustache typically has things like conditional `{{#foo}}` that will cause the JSON to not parse

**Pull request merged**